lib/lightning/work_orders.ex

```diff
@@ -434,70 +441,46 @@ defmodule Lightning.WorkOrders do
   @spec retry_many(
           [WorkOrder.t(), ...] | [RunStep.t(), ...],
           [work_order_option(), ...]
-        ) :: {:ok, count :: integer()} | UsageLimiting.error()
+        ) ::
+          {:ok, enqueued_count :: non_neg_integer(),
+           discarded_count :: non_neg_integer()}
+          | UsageLimiting.error()
+          | {:error, :enqueue_error}
   def retry_many([%WorkOrder{} | _rest] = workorders, opts) do
-    attrs = Map.new(opts)
-    orders_ids = Enum.map(workorders, & &1.id)
-
-    run_numbers_query =
-      from(att in Run,
-        where: att.work_order_id in ^orders_ids,
-        select: %{
-          id: att.id,
-          row_num:
-            row_number()
-            |> over(
-              partition_by: att.work_order_id,
-              order_by: coalesce(att.started_at, att.inserted_at)
-            )
-        }
-      )
-
-    first_runs_query =
-      from(att in Run,
-        join: attn in subquery(run_numbers_query),
-        on: att.id == attn.id,
-        join: wo in assoc(att, :work_order),
-        where: attn.row_num == 1,
-        order_by: [asc: wo.inserted_at],
-        preload: [
-          :dataclip,
-          :starting_job,
-          work_order: wo,
-          starting_trigger: [edges: :target_job]
-        ]
-      )
-
-    runs = Repo.all(first_runs_query)
+    runs_ids =
+      workorders
+      |> Enum.map(& &1.id)
+      |> get_first_runs_for_retry()
 
     with project_id <- Keyword.fetch!(opts, :project_id),
          :ok <-
            UsageLimiter.limit_action(
-             %Action{type: :new_run, amount: length(runs)},
+             %Action{type: :new_run, amount: length(runs_ids)},
              %Context{
                project_id: project_id
              }
            ) do
-      results =
-        Enum.map(runs, fn run ->
-          starting_job =
-            if job = run.starting_job do
-              job
-            else
-              [edge] = run.starting_trigger.edges
-              edge.target_job
-            end
-
-          do_retry(
-            run.work_order,
-            run.dataclip,
-            starting_job,
-            [],
-            attrs[:created_by]
-          )
-        end)
+      creating_user = Keyword.fetch!(opts, :created_by)
+
+      runs_ids
+      |> Enum.chunk_every(@run_many_chunk_size)
+      |> Enum.map(
+        # TODO: Enqueue the first chunk directly with Runs.enqueue_many and schedule the rest with Oban
+        &RetryManyWorkOrdersJob.new(%{runs_ids: &1, created_by: creating_user.id})
+      )
+      |> then(fn jobs ->
+        Oban.insert_all(Lightning.Oban, jobs)
+      end)
+      |> case do
+        inserted_list when is_list(inserted_list) ->
+          enqueued_count = length(runs_ids)
+          discarded_count = length(workorders) - enqueued_count
 
-      {:ok, Enum.count(results, fn result -> match?({:ok, _}, result) end)}
+          {:ok, enqueued_count, discarded_count}
+
```

Comment:
@josephjclark so caters for when the WorkOrder is rejected right, when the number of runs enqueued is less that the number of workorders that was selected?